### ISSUE TITLE
Suggest features you would like to see here

### ISSUE THREAD

**Issue by roguelike2d** [roguelike2d/TekkenBot#18]
No, I am not making it talk like ToolAssisted's does.

**Comment by senseirain** [comment]
The Overlay is Superb!,.. Yep More Like ToolAssisted's Vocals and bots,..

**Comment by WAZAAAAA0** [comment]
My 2 main suggestions:

1. Implement what I call the "raw method" to calculate frame advantage correctly 100% of the times as an alternative "fallback method" to all the convoluted calculations that you are currently trying to implement, for the people who manually decide to use it. Maybe add a config to enable it.
This is how the raw method works:
Just do P1_frame_count (14334B4A4) MINUS P2_frame_count (14334E294) and viceversa, as simple as that.
This method worked perfectly in my DOA5LR Frame Counter tool with no additional trickery, but Tekken has IASA frames, so the player has to manually hit/make the CPU block an attack and then make both characters Jump every single time.
This method requires more manual inputs by the user, but the calculated frame data should be 100% accurate in every situation, even after all sorts of moves with weird stance properties, delayed air attacks, attacks hitting on a later active frame and so forth and so on.
Practical frame counter example tool for current patch: http://www.mediafire.com/file/cxamccla6wquj3d/TopKekken.zip

2. Start distributing the Cheat Engine cheat table file with all the addresses you use, either within the source code, in the releases, or both. Please make our life easier!

**Comment by roguelike2d** [comment]
1. This is an idea that vaguely crossed my mind at points and now that I know there is some frame data that I'm not going to be able to calculate correctly it's probably a good idea. The difficulty is in directing users to it. While it's natural for me to set the dummy to jump from playing 2D fighters, I notice a most experienced Tekken players test frame data in other ways, so there's a double barrier of letting people know that its the fallback and introducing the idea that this is a way to test frame data in the first place.

1. Added a cheat table to repo for anyone who wants to use it.

**Comment by compsyguy** [comment]
It would be nice to have all the commands available in the bot gameplan and be able to pull moves by their block frames.

**Comment by roguelike2d** [comment]
I'm gonna wait and see what @dantarion does on that front. I'm hoping he publishes some of his work on the state engine at some point and I can make use of that. The other thing that'd be useful to track is some sort 'how far does this move tend to reach' stat so a bot can determine if a move or punish will reach.

**Comment by blankaex** [comment]
Tbh I'm more interested in the ArtificialKeyboard and GameInputter. Are these ready to be compiled and run? No GUI is fine.

**Comment by roguelike2d** [comment]
Yeah, they work fine. I'll probably release something in a week or so or if you download the source code, you can use the ugliest GUI in the world with GUI_FrameTrapBot.py to see it in action. It uses the p2 keyboard controls so the easiest way to use it is to go to practice, set the mode to 'controller' hit start on your controller and then hit HOME to give p2 the p2 keyboard controls. FrameTrapBot will then perform its given notation whenever it gets out of block stun (so jab it to get it to play).

Current notation, which is subject to change, is a comma seperated list of commands designed to be as similar to common Tekken notation as possible. Here's some notation examples for anyone wanting to try FrameTrapBot:

|   Example   |   Info     |
|---------|---------------------|
| +4                         | buffers a magic 4 before getting out of block stun.                             |
| 4, +1                     | waits 4 frames then presses 1                                                               |
| +1, 2, b+1+2        | taps button 1, waits two frames, taps back 1+2                                   |
| d | buffers a sidestep out of block stun by tapping down |
| u, 16, +3 | sidesteps for 16 frames (little less actually because of the buffer), then presses 3 |
| >                           | A 'smart wait' until either the next move or the next time recovery ends     |
| +1, >, +1, >, +2    | The Mishima punch string                                                                    |
| d/f+1, >>, d/f+1   | Two down forward ones                                                                       |
| S! | A smart wait with a bit of delay so that wr, ff, and other input moves come out properly |
| u/f+4, >>, 1, qcb+4, S!, qcf, 10, +2 | A basic Paul combo consisting of a launcher, a smart wait for the next move, a 1 frame wait so that qcb buffers properly, a screw attack, waiting for the attack to end, a qcf motion, waiting 10 frames for the dash to reach, and finally inputting the +2 to get deathfist|


Note that frame bot prepends a '>' to the string so that the string will come out on block stun. 

You can see more (possibly outdated) notation examples in the CharacterData folder.

**Comment by blankaex** [comment]
Thanks for the reply and all the info! Figured most of it from the brilliant documentation but it's good to see it all written out nicely. Are all the necessary files to run it included in the repo, though? I cloned it and tried to run it myself but this is what I'm getting.

```
A GUI that launches the frame trap bot, to practice frame traps on. The move the bot does out of block can be changed.

: not found
: not foundTrapBot.py: 5: ./GUI_FrameTrapBot.py:
: not foundTrapBot.py: 6: ./GUI_FrameTrapBot.py:
from: can't read /var/mail/tkinter
from: can't read /var/mail/tkinter.ttk
from: can't read /var/mail/_TekkenBotLauncher
from: can't read /var/mail/BotFrameTrap
./GUI_FrameTrapBot.py: 11: ./GUI_FrameTrapBot.py: import: not found
: not foundTrapBot.py: 12: ./GUI_FrameTrapBot.py:
: not foundTrapBot.py: 13: ./GUI_FrameTrapBot.py:
./GUI_FrameTrapBot.py: 14: ./GUI_FrameTrapBot.py: Syntax error: "(" unexpected
```

**Comment by roguelike2d** [comment]
That looks like the error you get when you try to run a python script as a shell command, you need to run it as 'Python GUI_FrameTrapBoty.py', But not with the latest version of python because Python 3.6 has a bug that stops TekkenBot from working (or more likely, the other way around). Anyway, use Python 3.4 or 3.5 for best results.

**Comment by omnislash79** [comment]
the ability to move the overlay instead of just 3/4 viewable at the top.

**Comment by roguelike2d** [comment]
There's a window mode option in TekkenData/frame_data_overlay.config that lets you drag it around.

**Comment by lovegu** [comment]
It would be nice if there's a got low parried state

**Comment by roguelike2d** [comment]
Wait, are there lows that don't get low parried? If anyone can give me some moves, I'll see if I can find the property.

**Comment by compsyguy** [comment]
Yes there are a few lows that can't get low parried... I know Yoshimitsu's unblockable low sword sweep can't be low parried... 
I think the bear's low spin move can't be either...

**Comment by yonecessito** [comment]
Asuka's standing 2 is listed as a 13 frame startup, but its actually 12.

**Comment by lovegu** [comment]
Yoshimitsu without sword, u1+2 and BT d+1 with/without sword cannot be parried

**Comment by AEEEE15** [comment]
You retard tekken bot does not work.

**Comment by dimesc** [comment]
Could you possibly incorporate which break is needed when thrown? Or even the input that was used for the throw.

**Comment by roguelike2d** [comment]
Found the variable that looks like the throw break byte, I'll add it to the notes section in the next release.

**Comment by lovegu** [comment]
When being thrown or while in breaking throw animation, some throws(king's 3+4,1+4 chain grap starter, havent tested others yet) makes bot.simple_state as SimpleMoveStates.GROUND_FACEUP

**Comment by noodalls** [comment]
Would it be possible to list the total frames for a move. The reason for asking for this is in making combos, often you can get two moves to clash. The results in replacing your normal recovery with the opponent's move's hit recovery, which can open up some interesting situation (e.g. Nina has a move with very quick recovery which was ideal for linking two moves that wouldn't normally work). 

However, the only way to know how this works is by knowing the total frames for a move (not the usual relative recovery between the two players.) 

There are also some instances where moves total frames change depending on the block/hit/CH/whiff state, which is interesting in and of itself. 

I hope you can do this, if not great work and I am thoroughly amazed by the progress of this project, the data we have now is so far ahead of previous Tekken games.

**Comment by roguelike2d** [comment]
Yeah I can add total move frames, block/hit stun frames, and attack recovery frames pretty easily. I've got all those kicking around. Total move frames is a little buggy at the moment but it should work for most moves.

@lovegu some of the simple state stuff is more reliable than others, all the labels in MoveInfoEnums.py are mostly guesstimation and at this point I wouldn't rely on them being real accurate.

**Comment by illua0** [comment]
I'd definitely like to see fullscreen support. unfortunately my laptop sucks so borderless wont work for me

**Comment by biejis** [comment]
recovery is unnecessary, you never will see a faster trade to continue a combo and you can calculate "start+active+recovery=total", and block, hit and counter hit are the same on main display

**Comment by biejis** [comment]
may a good idea for the transparent fuzzy display,  put it over the lifebar or simply rebuild the current items like this ex. with gigas on the screen
![fd6](https://user-images.githubusercontent.com/29903063/28249270-55547f36-6a20-11e7-9fae-7dc15e657fb3.jpg)

**Comment by roguelike2d** [comment]
@illua0 fullscreen is very far away right now. It requires modifying the game's dll and hooking into the directx calls that draw the game which is a bunch of work that I'm not sure I want to do anyway since not modifying the dll is a nice feature. However, that means it's probably not gonna work on single screen laptops that can't run in windowed mode for the indefinite future. 

About the only hope I can offer is that there's folks making Tekken 7 run on laptops it shouldn't run on by modifying the config files until it looks like a ps1 game. Some of those tweaks might let you get 60 fps windowed.

@biejis I debated a little whether or not total frames was useful enough to include, but ultimately I thought it was informative to see it there without doing a bunch of addition. Also exposing total frames can help identify cases where the calculation is wrong, since startup, active, and recovery frames are all calculated but total frames is just pulled directly from the game's memory.

As far as gui redesign goes, at this point, I think it's easier to stick with the current one that expands fairly easily. Once I've nailed down everything I want to include or perhaps provide a way for users to customize the columns themselves, I'll consider something that uses the space more efficiently.

**Comment by noodalls** [comment]
@biejis 

https://www.youtube.com/watch?v=s55S9cfLNEI
https://www.youtube.com/watch?v=2BrtcLWFZq4
https://www.youtube.com/watch?v=xRQP-6sKP0g

These are all combos that don't normally work, but depend on shortening your recovery with specific moves on hit. I think there has also been a system change where two moves which clash will now both CH, which might require sorting out which moves still work well for this and which no longer do. This is why I am requesting either total frames or recovery frames.

**Comment by lovegu** [comment]
I've found offset of Player HP(0x1B148), and the timer(0x4D9B8)

player hp is in percent 0~100

timer value is seconds*60

**Comment by roguelike2d** [comment]
@noodalls latest release (0.3.2) has both total frames and recovery if you want to try it.

@lovegu added the hp percent. The timer value is the same as the one at 0x19AD0. The game stores 8 'frames' of player data, presumably for rollback. Each one is 0x19F70 bytes so offsets higher than that belong to other frames in the sequence.

**Comment by noodalls** [comment]
Amazing, thank-you.

**Comment by dimesc** [comment]
Could you reduce the amount of "No fight detected. Gamestate not updated." commands displayed?

**Comment by lovegu** [comment]
move id for low parry:
MoveId - parried button
219 - 4
221 - 3
223 - 1
225 - 2

also when gets low parried move id changes to the values above +1
as an example when d4 gets parried move id changes to 220

**Comment by roguelike2d** [comment]
Yeah, the display is messy enough now that I'll put that in soon.

**Comment by biejis** [comment]
great work, help me pls,  which file have the code to edit the display windows, and which program edit this .pyd, i want to try test other display boxes. :SSSSS, i have anaconda 3

**Comment by roguelike2d** [comment]
GUI_TekkenBotPrime.py and GUI_FrameDataOverlay.py contain the manager window and the overlay. If you have python (64 bit only, 3.4 or higher), you can download the source code and run them with "python GUI_TekkenBotPrime.py". The .pyd is generated by pyinstaller, I have no idea how to edit it directly.

**Comment by biejis** [comment]
Help, I'm trying to combos just frame moves  and my old macro application has problems in fps, someone have a new and good DDDDDD

**Comment by roguelike2d** [comment]
Please use this issue to suggest features you would like to see. Not to ask for 3rd party macro applications.

**Comment by biejis** [comment]
i know, sry , but i don t have more people researching

**Comment by lovegu** [comment]
I am currently working on finding the values of the matched player while waiting in 'Warm up' mode.
so far, I've find out there's two ways of finding matches
first type is, the matching system searches from 'match list' which is now hidden and selects randomly one(usually the first one from the match list) from there and request for a match

the second type is the other way, other player finds me from the match list and request for a match

I have found the addresses for match list(charid, player name, wins, ranks, ping)
but for warm up I've only found player name and wins.

here's the cheat engine table data for it, hopefully it is useful for some people


Match list(next list item can be found by adding 218 to the first offset)
`    <CheatEntry>
      <ID>1138</ID>
      <Description>"Matchlist0_PlayerName"</Description>
      <LastState RealAddress="451A1698"/>
      <VariableType>String</VariableType>
      <Length>48</Length>
      <Unicode>0</Unicode>
      <CodePage>0</CodePage>
      <ZeroTerminate>1</ZeroTerminate>
      <Address>"TekkenGame-Win64-Shipping.exe"+03336410</Address>
      <Offsets>
        <Offset>138</Offset>
        <Offset>2C0</Offset>
      </Offsets>
    </CheatEntry>
    <CheatEntry>
      <ID>1205</ID>
      <Description>"Matchlist0_PING"</Description>
      <LastState Value="5" RealAddress="451A1674"/>
      <VariableType>4 Bytes</VariableType>
      <Address>"TekkenGame-Win64-Shipping.exe"+03336410</Address>
      <Offsets>
        <Offset>114</Offset>
        <Offset>2C0</Offset>
      </Offsets>
    </CheatEntry>
    <CheatEntry>
      <ID>1137</ID>
      <Description>"Matchlist0_CharId"</Description>
      <LastState Value="3" RealAddress="451A16E0"/>
      <VariableType>4 Bytes</VariableType>
      <Address>"TekkenGame-Win64-Shipping.exe"+03336410</Address>
      <Offsets>
        <Offset>180</Offset>
        <Offset>2C0</Offset>
      </Offsets>
    </CheatEntry>
    <CheatEntry>
      <ID>1141</ID>
      <Description>"Matchlist0_Rank"</Description>
      <LastState Value="21" RealAddress="451A16E4"/>
      <VariableType>4 Bytes</VariableType>
      <Address>"TekkenGame-Win64-Shipping.exe"+03336410</Address>
      <Offsets>
        <Offset>184</Offset>
        <Offset>2C0</Offset>
      </Offsets>
    </CheatEntry>
    <CheatEntry>
      <ID>1140</ID>
      <Description>"Matchlist0_Wins"</Description>
      <LastState Value="93" RealAddress="451A16E8"/>
      <VariableType>4 Bytes</VariableType>
      <Address>"TekkenGame-Win64-Shipping.exe"+03336410</Address>
      <Offsets>
        <Offset>188</Offset>
        <Offset>2C0</Offset>
      </Offsets>
    </CheatEntry>
    <CheatEntry>`

Warm up
`<CheatEntry>
      <ID>1197</ID>
      <Description>"WARMUP PLAYER NAME1"</Description>
      <LastState RealAddress="52D30CDC"/>
      <VariableType>String</VariableType>
      <Length>45</Length>
      <Unicode>0</Unicode>
      <CodePage>0</CodePage>
      <ZeroTerminate>1</ZeroTerminate>
      <Address>"TekkenGame-Win64-Shipping.exe"+0333B8A0</Address>
      <Offsets>
        <Offset>0</Offset>
        <Offset>50</Offset>
      </Offsets>
    </CheatEntry>
    <CheatEntry>
      <ID>1214</ID>
      <Description>"WARMUP PLAYER WINS1"</Description>
      <LastState Value="32" RealAddress="52D30CA8"/>
      <VariableType>4 Bytes</VariableType>
      <Address>"TekkenGame-Win64-Shipping.exe"+0333B8A0</Address>
      <Offsets>
        <Offset>-34</Offset>
        <Offset>50</Offset>
      </Offsets>
    </CheatEntry>
    <CheatEntry>
      <ID>1211</ID>
      <Description>"WARMUP PLAYER NAME2"</Description>
      <LastState RealAddress="52D30E1C"/>
      <VariableType>String</VariableType>
      <Length>45</Length>
      <Unicode>0</Unicode>
      <CodePage>0</CodePage>
      <ZeroTerminate>1</ZeroTerminate>
      <Address>"TekkenGame-Win64-Shipping.exe"+0333B8A0</Address>
      <Offsets>
        <Offset>140</Offset>
        <Offset>50</Offset>
      </Offsets>
    </CheatEntry>
    <CheatEntry>
      <ID>1213</ID>
      <Description>"WARMUP PLAYER WINS2"</Description>
      <LastState Value="58" RealAddress="52D30DE8"/>
      <VariableType>4 Bytes</VariableType>
      <Address>"TekkenGame-Win64-Shipping.exe"+0333B8A0</Address>
      <Offsets>
        <Offset>10c</Offset>
        <Offset>50</Offset>
      </Offsets>
    </CheatEntry>`

**Comment by roguelike2d** [comment]
I've added these values to the Cheat Engine Table. The table was getting a little overloaded so it's split into match and non-match values for now.

**Comment by eedahl** [comment]
One thing that would be really useful would be if the app could dump frame data to a json file. I'm making a move browser right now and would find it invaluable to augment and correct the data.

**Comment by roguelike2d** [comment]
Would you still want that knowing that there would no way to identify the move? Like, the app doesn't know that something is a 1,1 or a 1,2 or a qcf+1, all it has is a unique identifier, essentially a number. So I can give you a unique identifier for every move, but you'll have to manually map those to strings.

The unique identifiers don't always play nice with the frame data display (for example, let's say you cancel into a stance, depending on whether that stance cancel happens before or after the move is recorded in the log, you might get the move_id of the original move or of the stance cancel), but I think that can be worked around.

Having that map would be really useful though, it'd let the app display much more human input data.

**Comment by compsyguy** [comment]
Is there a way to tell if a character is in a stance? I took a quick look through the code and didn't see anything?

Also, is there a way to have the program send the 'reset position' command?

**Comment by eedahl** [comment]
I'd still be interested, though I couldn't make any promises as far as making use of it right away goes, since I'm still early in development. In other words, if you can find a use for it for yourself that would be cool, and maybe the whatever kinks remain can be ironed out.

I expect over 6500 identifiers, so... well there may be a way to do it but I fear headaches. That said, even partial data could be useful.

**Comment by roguelike2d** [comment]
@compsyguy I haven't found any stance stuff. In a sense, every string is a 'stance'. There's a bot command (BasicCommands.BotCommands.ResetPractice()) that presses start+accept and resets practice.

@eedahl I'll see about adding move_id back in and try to make it as consistent as I can. Printing to a file should be straightforward.

**Comment by eedahl** [comment]
Cool, could turn out to be very useful. I personally wouldn't need duplicates in the file, but handling duplicates isn't exactly an issue either so.

**Comment by eatsports** [comment]
any chance for an auto-updater for the overlay?

**Comment by roguelike2d** [comment]
I'm not sure I'm comfortable setting up a secure auto-updater and I'd kinda hate for TekkenBot to launch the a ransomware wave. I also have very little desire to manage migrations of user settings and data. I can probably add a nag pop-up when a new version is released though along with what features/bug fixes are in the release,

**Comment by RookieLongTIme** [comment]
Thank you for the great program. My request is to show on-block numbers on hit; currently they are marked with ?? Thank you!!

**Comment by roguelike2d** [comment]
Yeah, that'd be nice. There are a couple of issues between that happening and right now though.

1. The frame data is read live so it has to be generated first. You have to actually block a move before the data for it appears. This could be alleviated by putting it in a file so it'd 'remember' between launches.
2. Depending on whether you canceled the move into another move that 'block' value might be either the correct on block number or a whiff number. So if you press just 1, it'll show +1 or whatever for your jab, but if you do 1, 2, it'll show -15 as if the second jab had whiffed. This is actually intentional behavior to get multihitting moves to show correctly, but it means there are multiple on block values to show.
4. There's still a bunch of garbage values that show up in the frame data display for things like walls, ground bounces, throws and what not. I don't really want to record or display those but there's no way to tell the garbage data from the good data, if there was, it wouldn't be getting displayed in the first place.
3. Active frames have to be accounted for.

**Comment by jonhammer** [comment]
@roguelike2d 

Is it possible to pull specific match data and store it in some kind of log file for analysis?

My thoughts are it would be useful to have data on things like:
- Moves that hit me
- Strings that hit me
- Strings where I block the first hit and am hit by a subsequent hit
- Strings with high hits that I fail to duck
- Moves that I block and fail to punish
- Character specific win/loss ratios

Using this data I could guide my training and focus on particular weak spots. 

I understand certain unique move properties might throw off this data a bit, but perhaps there would be enough good data to still make this viable and worthwhile. 

If this if outside of the scope of the project I understand, but possibly you could point me roughly in the direction of where this data could be accessed and I can try to whip something up.

**Comment by roguelike2d** [comment]
Yeah, you may have noticed the primordial ooze for some of that stuff is already in the log.  Since a version or two ago it prints things like`p2:  LOW | 7 | 1 | 3600 | 3600 | HIT`  or `p2:  WHIFF_PUNISH | 22 | 2 | 3600 | 3600 | HIT`. Basically it tries to explain why you got hit.

It's... kinda interesting so far. Not as useful as I had hoped. Really to be useful the presentation needs to be super slick since during the match it's basically impossible to read and after the match it's not really better than just spending 10 seconds thinking about about how you got hit.

Also since last version it tracks your w/l record vs characters and specific steam users if you leave it running during ranked matches, you can see it by switching the mode to MatchupRecord.

Failing to duck high strings and strings that successfully frame trap is interesting because that's information you really want to be learning as the match is happening. While practicing punishes for a specific move is something you want to know you have to do after the match is over.

**Comment by jonhammer** [comment]
> it's not really better than just spending 10 seconds thinking about about how you got hit.

I have to disagree with that. Usually when someone plays ranked they are playing many matches in a row. You might think about what moves are hitting you, but that isn't nearly as valuable (or reliable) as having a bunch of historical data. I'm not going into practice mode between every match so if I get hit by a move early on and don't write it down or something, I'm not going to remember to get the reps in in my next practice mode session.

There is also the benefit of having potentially hundreds or even thousands of matches worth of data. Maybe you think you know a certain punish, but when you analyze the data, you realize you are only punishing it 50% of the time or something. Or perhaps you are using suboptimal punishes. Ultimately I think there could be many other advantages I haven't even considered if you took a sort of data analytics approach. I'm imagining something like pushing all the data into splunk or some similar tool and being able to analyze and correlate it. You could have a list of the top 10 moves you suck at punishing, or top 10 strings by character you fail to block/duck, and you could identify if you had more trouble punishing while playing as certain characters, etc. You could track how much impact your practicing blocking/ducking certain moves has on your in-match efficiency over time.

> Failing to duck high strings and strings that successfully frame trap is interesting because that's information you really want to be learning as the match is happening. While practicing punishes for a specific move is something you want to know you have to do after the match is over.

I think failing to duck a string that has a high hit (which will often result in a ws launch punish) is roughly equivalent to failing to punish a move on-block and something that needs to be identified (in my imagined world here, via analyzing the data) and then taken into practice mode.

tl;dr: Players should take note of moves that give them problems and practice against them, but having the data could prove to be a very useful tool in terms of identifying this better than making mental notes.

**Comment by zyns** [comment]
> Would you still want that knowing that there would no way to identify the move? Like, the app doesn't know that something is a 1,1 or a 1,2 or a qcf+1, all it has is a unique identifier, essentially a number. So I can give you a unique identifier for every move, but you'll have to manually map those to strings.

@roguelike2d I'd definitely be interested in this one and helping you out with it, as it should be mostly just manual grunt work. There should be several applications for it. DoA had a good system with this, as it displayed the possible options from a string for learning purposes in local play. Tekken needs something similar to make the learning more efficient.

Building up a list of moves and their properties just seems like it would be damn useful. I'd just love to learn to dodge high strings (and punish them) without constantly flipping to practice mode and to different websites, making the whole process just take 10x longer for no good reason.

**Comment by JABirchall** [comment]
I would personally like to see it as an AI, possibly using https://www.tensorflow.org/

Train it the moves and watch it learn.

**Comment by roguelike2d** [comment]
@JABirchall I talked a little about the AI issues at #35 and haven't been working on it much. It might be worthwhile to ignore future roadblocks and just try to get something working though.

@zyns I'm going to look for a way to do it algorithmic ally for a bit longer, then I'm probably gonna appeal to the community (google docs spreadsheet maybe?). The good news is that unique identifiers have fairly readable strings associated with them now now in the dev version (strings like 'cr2st_00' for crouching to standing or 'Cz_lp00' for jab). It's unclear if I can generate a human useful movelist with standard names (1, 1, 2) at this point.

@jonhammer I think that stuff is very achievable, but it might require some human input on what strings are punishible and how punishible they are, factoring things like pushback and possible move cancels.

**Comment by JABirchall** [comment]
> @JABirchall I talked a little about the AI issues at #35 and haven't been working on it much. It might be worthwhile to ignore future roadblocks and just try to get something working though.

I just read that, and while some roadblocks may be an issue especially the good data part, need a good player to train the AI, unless the AI can learn from a bad player such as my self. But you do have an advantage over most which is reading a substantial amount of data from memory. Most people who take on an AI task just go by what they can see, like self driving AI. The ai would know more then someone watching would and at a faster rate.

I believe tensor flow works on windows, as I only found out about it from a GTA V self driving AI which uses it (and another library). the only roadblock I know of is AI/ML is very GPU intensive as that is the most efficient way to achieve ML. I have a 1080ti which could maybe handle it, but I am not a python dev, and never used ML libraries before. I can tinker.

**Comment by eedahl** [comment]
I'm pretty sure you canuse Tensorflow on any platform that supports Python

**Comment by excision1** [comment]
yeah i would love to see the ai/bot get some love been having a blast adding combos and such to the bot punisher/character txt files and swapped block all lows to parry all lows crazy stuff to fight against in practice  any chance of some form of whiff punisher possible ?

**Comment by roguelike2d** [comment]
I've had some issues getting a consistent measurement of range out of the memory. I've got hurtbox positions, so you'd think it'd be pretty easy but I haven't quite gotten a consistent way to measure range down. About the best I could do right now is give the bot a 'yes/no, was I able to punish {x} move at {y} range with {z} move'.

**Comment by excision1** [comment]
ahhh i see one thing i was gonna ask was u see on your punisher bot etc how to go about edit the moving 
 part of the bot and what side of screen it activates on as i like to play right side of screen but bot works right side so would pref it on left side i understand u said lil support been browsing the files for a while trying to test lil changes here n there boredom :P

**Comment by biejis** [comment]
crush column

c=		crouch
cb=		crouch back
cf=		crouch forward
csl=	        crouch side left
csr=	        crouch side right

cs=		crouch special (can avoid mid range)
csb=	crouch special back
csf=	        crouch special forward
cssl=	crouch special side left
cssr=	crouch special side right

a=		air
ab=		air back
af=		air forward
asl=	        air side left
asr=	        air side right

as=		air special (can avoid mid range)
asb=	air special back
asf= 	air special forward
assl=	air special side left
assr=	air special side right

l=		side left 
r=		side right



ex:		devil jin
		u4
		start / crush
		20-21 a+c9
		
		josie
		uf3
		start / crush
		28-29 asf9
		
		jin
		uf2
		start / crush
		15-16 csr8
		
		akuma
		ub
		start / crush
		8	   asb10

**Comment by biejis** [comment]
hurt/hit range viewer is useful?
https://images-ext-2.discordapp.net/external/CGyAOug_mjMu1Xuah-UXCagoKjjGoaK10Qg37rgko_0/https/pbs.twimg.com/media/DCGEcCNUAAAbK4o.jpg%3Alarge?width=1060&height=597

**Comment by JABirchall** [comment]
I suppose what tekkenbot would need to know if n attack would hit is the skeleton of both the players.

**Comment by cherubrock** [comment]
I would like FrameDataOverlay display text in some monotype (eg. courier) or configurable font via frame_data_overlay.ini

**Comment by roguelike2d** [comment]
The @toolassisted and @dantarion hurtbox viewers are, as I understand it, taking advantage of some known unreal engine stuff and injecting the dll to get the engine to draw those boxes. Both those guys got it working pretty quickly after the game released, but I have no idea how to do it and even if I did it's kinda at odds with TekkenBot's current "don't modify the .dll" ethos.

If I was super motivated and there was no other way, I can sorta imagine some avenues to explore. But I'm not and there is, so that's probably not gonna happen.

@cherubrock While looking into your request, I found a bug in the code that was preventing the monospaced font I'd set from working. So next release the bars should line up.

**Comment by JABirchall** [comment]
This is how they did it: https://github.com/KN4CK3R/UnrealEngineSDKGenerator

**Comment by B-ballRemovalDept** [comment]
The ability to add custom sounds for when you successfully punish a move. I don't know if this is already a feature but a way to change the ranges for what is launch punishable so if my character doesn't have a 14 frame launcher it doesn't show it as a launch punishable move,

**Comment by roguelike2d** [comment]
Whoa, someone used an overlay.

Successful punish sounds are easy enough to add, I can add a way to customize punish thresholds too. I might do something like base it on 'damage missed' using the list from http://www.avoidingthepuddle.com/news/2016/12/15/tekken-7-fated-retribution-character-punishment-rankings.html

**Comment by jasonxfong** [comment]
Hi, may I know how to resolve this error?
The framedatas just showing Question Marks.

Parsing movelist for 
Tekken pid acquired: 12248
Trying to acquire Tekken library in pid: 12248
Found TekkenGame-Win64-Shipping.exe
ReadProcessMemory Error: Code 299
ReadProcessMemory Error: Code 299
ReadProcessMemory Error: Code 299
ReadProcessMemory Error: Code 299
ReadProcessMemory Error: Code 299
ReadProcessMemory Error: Code 299
ReadProcessMemory Error: Code 299
ReadProcessMemory Error: Code 299
Getting Block of Data Error: Code 299
Fight detected. Updating gamestate.
ReadProcessMemory Error: Code 998
ReadProcessMemory Error: Code 299
ReadProcessMemory Error: Code 299
ReadProcessMemory Error: Code 998
ReadProcessMemory Error: Code 299
ReadProcessMemory Error: Code 299
Getting Block of Data Error: Code 299
Getting Block of Data Error: Code 998
Parsing movelist for 
Parsing movelist for

**Comment by roguelike2d** [comment]
Probably because of a modded or cracked dll, it doesn't work if you change where the memory addresses are. Also, this issue is for suggesting features to add to TekkenBot, please make a new issue if you're encountering a bug.

**Comment by B-ballRemovalDept** [comment]
@roguelike2d I was playing with the Tekken-bot coach yesterday I think that's a good idea especially since a lot of the time I would punish with a weaker 10 frame move when I could use a stronger 13 frame move.
Do you also think you can enable the sound effects for the standard frame-data overlay?

**Comment by roguelike2d** [comment]
The only difficulty is the standard overlay doesn't know which side of the screen you're on in online play. There's a couple of possibilies that pop to mind:
* Make the punish coach overlay more like the the frame data overlay, but only show opponents data. This sorta fulfills another need since once you play your own character enough, you don't really need to see your own frame data.
* Make the frame data overlay better at figuring out which side of the screen the players are on by tracking whether p1 side or p2 side was last picked and make the sound functionality an overlay setting, This is nice because you don't have to go switching to another overlay and it'll remember that you had it on when you restart. In the future, add an option to only show your opponents data while playing online.

I kinda tend toward splitting up the functionality so that there's one you run online and one you run in practice mode. With the potential in the future to further specialize them.

**Comment by B-ballRemovalDept** [comment]
Yeah, that sounds good the reason I asked for the standard overlay is that I like to record my matches and look back at them with all the frame data there like being able to see the start up and whether it's a high mid or low.

**Comment by excision1** [comment]
yeah what b-ball said having issues with random side switches maybe a feature to force tell it what side/character etc

**Comment by wierdthing** [comment]
Is there a way to play the sound for the jab/launch punishes as the move is blocked rather than when the punish window closes?
